cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -242,15 +252,35 @@ private void processMetrics() {
           }
         }
 
+        if (nextMessageId != null) {
+          metaKeyMap.put(topicIdMetaKey, nextMessageId);
+          lastMessageId = nextMessageId;
+        }
+
         if (records.isEmpty()) {
           LOG.trace("No metrics record to process.");
           return;
         }
 
-        persistRecords(records);
-        records.clear();
-        lastMessageId = nextMessageId;
-        persistMessageId(lastMessageId);
+        if (!persistingFlag.compareAndSet(false, true)) {
```

Comment:
add comments about why we do this. Like how `htable` is not-thread safe and how this helps to ensure we call metricstore add records from just one of the thread, etc.